<a href="https://colab.research.google.com/github/Mohit352-bit/Machine-Learning/blob/main/Fake_News_Detector.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [15]:
import numpy as np 
import pandas as pd 
import string
import nltk
import matplotlib.pyplot as plt
import seaborn as sns
from nltk.corpus import stopwords
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import BernoulliNB


%matplotlib inline 

In [ ]:
nltk.download('stopwords')
nltk.download('wordnet')
nltk.download('punkt')

In [2]:
df = pd.read_csv('FakeNewsData.csv'  , index_col = 'id')
df

title  \
id                                                         
0      House Dem Aide: We Didn’t Even See Comey’s Let...   
1      FLYNN: Hillary Clinton, Big Woman on Campus - ...   
2                      Why the Truth Might Get You Fired   
3      15 Civilians Killed In Single US Airstrike Hav...   
4      Iranian woman jailed for fictional unpublished...   
...                                                  ...   
20795  Rapper T.I.: Trump a ’Poster Child For White S...   
20796  N.F.L. Playoffs: Schedule, Matchups and Odds -...   
20797  Macy’s Is Said to Receive Takeover Approach by...   
20798  NATO, Russia To Hold Parallel Exercises In Bal...   
20799                          What Keeps the F-35 Alive   

                                          author  \
id                                                 
0                                  Darrell Lucus   
1                                Daniel J. Flynn   
2                             Consortiumnews.com   
3                                Jessica Purkiss   
4                                 Howard Portnoy   
...                                          ...   
20795                              Jerome Hudson   
20796                           Benjamin Hoffman   
20797  Michael J. de la Merced and Rachel Abrams   
20798                                Alex Ansary   
20799                              David Swanson   

                                                    text  label  
id                                                               
0      House Dem Aide: We Didn’t Even See Comey’s Let...      1  
1      Ever get the feeling your life circles the rou...      0  
2      Why the Truth Might Get You Fired October 29, ...      1  
3      Videos 15 Civilians Killed In Single US Airstr...      1  
4      Print \nAn Iranian woman has been sentenced to...      1  
...                                                  ...    ...  
20795  Rapper T. I. unloaded on black celebrities who...      0  
20796  When the Green Bay Packers lost to the Washing...      0  
20797  The Macy’s of today grew from the union of sev...      0  
20798  NATO, Russia To Hold Parallel Exercises In Bal...      1  
20799    David Swanson is an author, activist, journa...      1  

[20800 rows x 4 columns]

In [3]:
df1 = df.copy()

### Removing Null values

Since there is no  nan values in text column and only  we  need 'text' column for the prediction rest all columns are useless so we do not need to remove null values from them


In [4]:
df1.dropna(subset= ['text'] , inplace=True)

In [5]:
df1.isnull().sum()

title      558
author    1918
text         0
label        0
dtype: int64

### Removing stopwords and puntuations from the text

With the help of list comprehension I removed the stopwords from the statements of the data and then with the help of string.puntuation pakage I removed the punctuations from the statments as both stopwords and punctuations are useless for prediction.

In [6]:
def text_processing(str):
    df_nopunc = [x for x in str if x not in string.punctuation] #using stopwors with list comprehension
    df_join = ''.join(df_nopunc) #joining the word without the stopwords to form a sentence again
    return [word for word in df_join.split() if word not in stopwords.words("english")]
            

### Spliting the data for training and testing

Spiliting the data in ration of 70% of data for training and 30% of data for the testing of model

In [7]:
X = df1['text']
y = df1['label']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

In [8]:
from sklearn.pipeline import Pipeline

### Model building
Using PIPELINE for the applying countvectorizer ,  TF-IDF and ML model training process and to assemble several steps that can be cross-validated together while setting different parameters.


Use of Contvectorizer - 
I used countvectorizer to convert the text to numerical so that it can be understood by machine learning model and we have used it inside a pipline to automate the process instead fo writing it for every model

Use of IDF Transformer - 
I used tf-idf transformer for measuring how important a term is in the data I have used so that it can be used for prediction whether the news is fake or real with help of particular words

Machine Learning models - 
model in pipeline is used to apply different machine learning algorithm for the prediction.







### Using Decision Tree Classifier for prediction 



In [27]:
pipeline = Pipeline([
    ('nopunc' , CountVectorizer(analyzer=text_processing)), #convert text to numerical
    ("tf-idf" , TfidfTransformer()),# measuring how important a term is in the data
    ('model' , DecisionTreeClassifier(criterion= 'entropy',splitter='best' , max_depth = 25)) #ml algorithm for used for prediction
    
])

In [28]:
pipeline.fit(X_train , y_train)

Pipeline(steps=[('nopunc',
                 CountVectorizer(analyzer=<function text_processing at 0x0000017A2FDB6620>)),
                ('tf-idf', TfidfTransformer()),
                ('model',
                 DecisionTreeClassifier(criterion='entropy', max_depth=25))])

In [29]:
pred2 = pipeline.predict(X_test)

In [30]:
from sklearn.metrics import classification_report
print(classification_report(y_test , pred2))

              precision    recall  f1-score   support

           0       0.97      0.96      0.96      3141
           1       0.96      0.97      0.96      3088

    accuracy                           0.96      6229
   macro avg       0.96      0.96      0.96      6229
weighted avg       0.96      0.96      0.96      6229



### Using Random Forest Algorithm

In [11]:
pipeline = Pipeline([
    ('nopunc' , CountVectorizer(analyzer=text_processing)), #convert text to numerical
    ("tf-idf" , TfidfTransformer()),# measuring how important a term is in the data
    ('model' , RandomForestClassifier(n_estimators=120 , random_state = 42)) #ml algorithm for used for prediction
    
])

In [12]:
pipeline.fit(X_train , y_train)
pred2 = pipeline.predict(X_test)

In [13]:
from sklearn.metrics import classification_report
print(classification_report(y_test , pred2))

              precision    recall  f1-score   support

           0       0.95      0.97      0.96      3141
           1       0.97      0.95      0.96      3088

    accuracy                           0.96      6229
   macro avg       0.96      0.96      0.96      6229
weighted avg       0.96      0.96      0.96      6229



### Using Support Vector Classifier for prediction

In [ ]:
from sklearn import svm

clf = svm.SVC(kernel='linear') 

pipe = Pipeline([('vect', CountVectorizer()),
                 ('tfidf', TfidfTransformer()),
                 ('model', clf)])

model = pipe.fit(X_train, y_train)
prediction = model.predict(X_test)

In [ ]:
from sklearn.metrics import classification_report
print(classification_report(y_test , prediction))

              precision    recall  f1-score   support

           0       0.97      0.96      0.96      3141
           1       0.96      0.97      0.96      3088

    accuracy                           0.96      6229
   macro avg       0.96      0.96      0.96      6229
weighted avg       0.96      0.96      0.96      6229



In [16]:
pipeline = Pipeline([
    ('nopunc' , CountVectorizer(analyzer=text_processing)), #convert text to numerical
    ("tf-idf" , TfidfTransformer()),# measuring how important a term is in the data
    ('model' , BernoulliNB()) #ml algorithm for used for prediction
    
])

In [17]:
pipeline.fit(X_train , y_train)
pred2 = pipeline.predict(X_test)

In [18]:
from sklearn.metrics import classification_report
print(classification_report(y_test , pred2))

              precision    recall  f1-score   support

           0       0.88      0.76      0.82      3141
           1       0.79      0.90      0.84      3088

    accuracy                           0.83      6229
   macro avg       0.84      0.83      0.83      6229
weighted avg       0.84      0.83      0.83      6229



### Decision Tree , SVM and Random Forest almost have perfromed same except Naive Bayes and the best model can predict with 96% Accuracy which is quite good performance of our model 
